src/vivarium/framework/resource/manager.py

```diff
@@ -73,47 +74,52 @@ def sorted_nodes(self) -> list[ResourceGroup]:
     def setup(self, builder: Builder) -> None:
         self.logger = builder.logging.get_logger(self.name)
 
-    # TODO [MIC-5380]: Refactor add_resources for better type hinting
     def add_resources(
         self,
         resources: Iterable[str | Resource],
-        producer: Any,
         dependencies: Iterable[str | Resource],
+        initializer: Callable[[SimulantData], None] | None,
     ) -> None:
         """Adds managed resources to the resource pool.
 
         Parameters
         ----------
         resources
             The resources being added. A string represents a column resource.
-        producer
-            A method or object that will produce the resources.
         dependencies
             A list of resources that the producer requires. A string represents
             a column resource.
+        initializer
+            A method that will be called to initialize the resources. This is
+            called during population initialization.
 
         Raises
         ------
         ResourceError
             If a component has multiple resource producers for the ``column``
             resource type or there are multiple producers of the same resource.
         """
-        resource_group = self._get_resource_group(resources, producer, dependencies)
-
-        for resource in resource_group:
-            if resource in self._resource_group_map:
-                other_producer = self._resource_group_map[resource].producer
+        resource_group = self._get_resource_group(resources, dependencies, initializer)
+
+        for resource_id in resource_group:
+            if resource_id in self._resource_group_map:
+                other_resource = self._resource_group_map[resource_id]
+                if resource_group.is_initializer:
```

Comment:
This added logic seems like it is a bugfix? Were we previously able to have a non-initializing resource that would result in an unwanted raise?

Reply:
This is to do with the fact that now not all resources will have an initializer (whereas preciously they all have producers). I just had an idea about how to improve this by passing the component that I will try out and maybe put up in a new PR.